### PR TITLE
ldrb mla mov mul mvn

### PR DESCRIPTION
**Pull request by RadhikaG**
review pl0x

### CODE REVIEW

File: isa.c [hunk 1]
```diff
@@ -63,7 +68,8 @@ static void decode_and_exec(uint32_t instruction)
             }
         } else {
             if (get_bit(instruction, L_BIT) == 1) { // LDRB
-                // @RadhikaG has to do this
+                exec_LDRB(instruction);
+                // no ure mum does dis evry night
```
Comment (jailuthra) [inline isa.c:75]: wew so funny

File: isa.c [hunk 2]
```diff
@@ -78,9 +84,19 @@ static void decode_and_exec(uint32_t instruction)
             case OP_EOR: exec_EOR(instruction); break;
             case OP_ORR: exec_ORR(instruction); break;
             case OP_TST: exec_TST(instruction); break;
+            case OP_MOV: exec_MOV(instruction); break;
+            case OP_MVN: exec_MVN(instruction); break;
         }
     } else if (get_bits(instruction, 27, 24) == 0xf) { // SWI
         exec_SWI(instruction);
+    } else if (get_bits(instruction, 27, 24) == 0x0 &&
```
Comment (jailuthra) [inline isa.c:95]: yah so decoder will read these inside data processing instructions. if you want to fix it within this PR, reply "yay" else i'll do it later

File: isa.c [hunk 3]
```diff
@@ -296,4 +312,85 @@ static void exec_B(uint32_t instruction)
...
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        uint8_t data = mem_read_8(ld_str_addr_mode(curr_state, &(next_state), instruction));
+        uint32_t rd_id = get_bits(instruction, 15, 12);
+        next_state.regs[rd_id] = data; // casting uint8_t to uint32_t zeros top 3 bytes on its own; done to store byte to LSB of rd_id
+    }
+}
+
+// Multiply and Multiply-Accumulate Instructions
+// MUL and MLA produce the low 32 bits of a multiply.
+// TIL: The results of a signed multiply and an unsigned multiply
```
Comment (jailuthra) [inline isa.c:392]: vnice
share interesting links on telegram bby, thx

File: isa.c [hunk 4]
```diff
@@ -296,4 +312,85 @@ static void exec_B(uint32_t instruction)
...
+// differ only in the upper 32 bits; the low 32 bits remain the same for both.
+// So we can use the MUL and MLA ops unchanged for signed and unsigned operands.
+static void exec_MUL(uint32_t instruction)
+{
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
```
Comment (jailuthra) [inline isa.c]: remove condition_checks from all your instructions, we're doing these on decoder stage now

File: isa.c [hunk 5]
```diff
@@ -296,4 +312,85 @@ static void exec_B(uint32_t instruction)
...
+        }
+    }
+}
+
+static void exec_MVN(uint32_t instruction)
+{
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        struct ShifterOperand * shiftop;
+        shiftop = shifter_operand(curr_state, instruction);
+        uint32_t val = ~(shiftop->shifter_operand) & 0xFFFFFFFF; // bitwise negation promotes result to (int); bitwise and-ing done to prevent this
```
Comment (jailuthra) [inline isa.c]: nice

**Review by jailuthra — APPROVED** [review]
fix the condition checks and optionally the decoder thingy, and it's good for merge

**Comment by jailuthra** [comment]
closes issue #12

File: isa.c (hunk 2 above)
Comment (RadhikaG) [reply isa.c:95]: yay

File: isa.c (hunk 5 above)
Comment (jailuthra) [reply isa.c]: oh so this wasn't actually necessary here as shifter_operand output is 32bits, but yah still leave as is for clarity

File: isa.c [hunk 6]
```diff
@@ -77,7 +77,7 @@ static void decode_and_exec(uint32_t instruction)
                 exec_STRB(instruction);
             }
         }
-    } else if (get_bits(instruction, 27, 26) ==  0x0) {
+    } else if (get_bits(instruction, 27, 26) ==  0x0 && get_bit(instruction, 4) == 0) {
```
Comment (jailuthra) [inline isa.c:80]: there are three types of data instructions, this just covers the first one. but anyway i'm merging your pr, i'll fix this after merge

**Pull request merged**